Convert nws integration to component configuration

MatthewFlamm opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
Changes superseded by #34267

## Proposed change
This PR updates `nws` integration to a component and centralizes the data update calls to reduce requests with multiple entities that partially share data.

I also believe that this now makes this integration match silver on the quality scale.  Is this just a check box in the PR, or do I have to add it somewhere else, like the documentation?

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
nws:
  api_key: YOUR_API_KEY
```

## Additional information
This PR creates two entities per entry in configuration, one for hourly forecasts and one for day+night forecasts.  The central data class enables only a single observation data call for each set of entities.  The entities are named by the station name either configured or determined automatically.

I have the config flow mostly done, but this seemed like a smaller chunk for a PR first.

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/11951

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [x] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

MatthewFlamm committed:
convert nws to component configuration

MatthewFlamm committed:
add more debug logging

MatthewFlamm commented:
I'm seeing that the `DataUpdateCoordinator` class is now available, which was introduced at about the same time as this PR.  It seems to provide a lot of the centralized API polling logic used here, but I need to look into it.  It is possible that the separate API calls here prevents clean usage of that class.

MatthewFlamm commented:
One question for feedback on configuration.  The integration platform configuration currently allows the user to specify the `name`, but in the component configuration, as proposed, would create two entities for each entry in the configuration.  Allowing `name` doesn't really make sense to me anymore, and I think users can now rename entities in the UI.  

Should there still be an option to have to prepend a name to each entity during configuration?  For example, currently the configuration:

```yaml
nws:
  api_key: MY_EMAIL
  station: ABC
```

generates two entities named `ABC_daynight` and `ABC_hourly`.  Should I allow something like:

```yaml
nws:
  api_key: MY_EMAIL
  station: ABC
  name_prefix: vacation_home
```

to generate something like `vacation_home_ABC_daynight` or `vacation_home_daynight`.

MatthewFlamm committed:
remove assumed state

MatthewFlamm committed:
move initialization of data into init

MatthewFlamm commented:
I looked at the DataUpdateCoordinator class, but it doesn't seem to work well in this case.  Here, there is one API call common to two entities and a separate API call unique to each entity.

bdraco commented on homeassistant/components/nws/__init__.py:
```diff
@@ -1 +1,195 @@
+        return self.nws.forecast_hourly
+
+    async def async_update(self, now=None):
+        """Update all data."""
```
There is a bit of code duplication here.

How about breaking this out into a function that updates each bit of data and sets the status that you could call 3 times?

MatthewFlamm replied:
Good suggestion.

MatthewFlamm committed:
refactor update logic

MatthewFlamm committed:
use forecast success for checking available entities

bdraco commented on homeassistant/components/nws/__init__.py:
```diff
@@ -1 +1,186 @@
+
+def unique_id(latitude, longitude):
+    """Return unique id for entries in configuration."""
+    return f"{DOMAIN}_{latitude}_{longitude}"
```
AFAICT, unique ids are per platform.   You should be able to drop DOMAIN



`helpers/entity.py`

```

    def __eq__(self, other):
        """Return the comparison."""
        if not isinstance(other, self.__class__):
            return False

        # Can only decide equality if both have a unique id
        if self.unique_id is None or other.unique_id is None:
            return False

        # Ensure they belong to the same platform
        if self.platform is not None or other.platform is not None:
            if self.platform is None or other.platform is None:
                return False

            if self.platform.platform != other.platform.platform:
                return False

        return self.unique_id == other.unique_id
```

MartinHjelmare replied:
Yes, the entity registry, which handles the entity unique ids, is aware of both the integration domain and platform domain of an entity.

MatthewFlamm replied:
Thanks for the detailed info.  I was under the impression that I had to provide a totally unique id across the entire ecosystem, but this makes much more sense.  I was wondering how I would handle the addition of more platforms, but this makes it clear to me.

bdraco commented on homeassistant/components/nws/__init__.py:
```diff
@@ -1 +1,186 @@
+            self.update_forecast_hourly_success,
+        )
+
+        async_dispatcher_send(self.hass, unique_id(self.latitude, self.longitude))
```
It looks like you are using the unique id here as well. I'm not sure dispatcher is unique to the domain so you may need to keep it here.

@balloob will know for sure.

balloob replied:
dispatcher is global so should keep using `DOMAIN`.

bdraco commented on homeassistant/components/nws/__init__.py:
```diff
@@ -140,56 +140,47 @@ def forecast_hourly(self):
-        try:
-            _LOGGER.debug("Updating forecast hourly for station %s", self.station)
-            await self.nws.update_forecast_hourly()
+        await self._async_update_item(
```
Nice work!  This is much easier to follow now

MatthewFlamm committed:
Split unique_id into more usable pieces.

use a base_unique_id for each entry.  Add domain for signaling to entities.  Each entity in each platform can use base_unique_id to form individual unique_id's.

bdraco commented:
> One question for feedback on configuration. The integration platform configuration currently allows the user to specify the `name`, but in the component configuration, as proposed, would create two entities for each entry in the configuration. Allowing `name` doesn't really make sense to me anymore, and I think users can now rename entities in the UI.
> 

Ideally, you are providing the name and they can customize on the frontend if they so desire.  It makes sense to remove `name`

bdraco commented on homeassistant/components/nws/weather.py:
```diff
@@ -165,85 +68,62 @@ def convert_condition(time, weather):
-        _LOGGER.error(
-            "Error getting station list for %s: %s", (latitude, longitude), status
-        )
-        raise PlatformNotReady
```
Removing `PlatformNotReady` means this integration will never retry if nws is unreachable without a restart.

Consider moving this to a config entry via a config flow https://developers.home-assistant.io/docs/config_entries_config_flow_handler/ in a **new** PR after this one is done.

MatthewFlamm replied:
This was the plan, and I think the reason why I had moved it out of the platform.  Would it be advisable to move this piece back here in the meantime?

bdraco replied:
As long as you are going to get the follow up PR in time before 0.108 ships it can wait for the next PR

MatthewFlamm replied:
I reverted this to happen inside `async_setup_platform`.  I can still try to have a PR for config flow by 0.108, but it is good to not have to rely on it.

MatthewFlamm committed:
Revert "move initialization of data into init"

This reverts commit 09eb0220469285b10f0500f5f6def67415931a81.

MatthewFlamm committed:
add silver quality scale to manifest

MatthewFlamm commented:
I rebased on dev since this was based on a commit from two months ago.  I also added this to silver quality scale in manifest, which answers one of my questions above, but obviously this is still up for review.

bdraco commented on homeassistant/components/nws/weather.py:
```diff
@@ -165,85 +73,71 @@ def convert_condition(time, weather):
-            return
+    async def async_added_to_hass(self) -> None:
+        """Set up a listener and load data."""
+        async_dispatcher_connect(
```
Please undo the dispatcher in 
```async def async_will_remove_from_hass(self):```

```suggestion
        self._dispatcher_undo = async_dispatcher_connect(
```

bdraco replied:
Or better wrap with `self.async_on_remove(`

bdraco commented:
@MatthewFlamm This is looking good.  What are your thoughts about converting this to a config flow so you can import the old configuration and avoid making this a breaking change?


```
python3 -m script.scaffold config_flow
```

MatthewFlamm commented:
I had a config flow written but it was about double or more the PR size with it in, so I thought I would make a smaller PR in hopes it would be easier to review and be reviewed faster.  I suppose not.

Won't it still be a breaking change since the current configuration is under the weather component?  Can you forward a platform config to a config flow?  

I don't think it will be a breaking change to add a config flow after this one?  If it will be, I agree that it probably should be done now.

MatthewFlamm commented:
@bdraco Looking forward, the one piece I haven't looked into very much is that the current scan_interval configuration option is probably better in an options flow, can the configuration of configuration.yaml be used to set options as well?  I thought I would bridge this during the config flow PR, but here we are.

bdraco commented:
> I had a config flow written but it was about double or more the PR size with it in, so I thought I would make a smaller PR in hopes it would be easier to review and be reviewed faster. I suppose not.

I'm all for smaller PRs but avoiding breaking changes is usually more important as long as we keep the PR to a manageable size.
> 
> Won't it still be a breaking change since the current configuration is under the weather component? Can you forward a platform config to a config flow?

Your existing `async_setup_platform` can be converted to a `async_setup_entry` and then replace `async_setup_platform` with something like this to allow it to be imported without having to alter the yaml config.

```
async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
    """Set up the nws platform."""

    hass.async_create_task(
        hass.config_entries.flow.async_init(
            DOMAIN, context={"source": SOURCE_IMPORT}, data=config
        )
    )
```
> 
> I don't think it will be a breaking change to add a config flow after this one? If it will be, I agree that it probably should be done now.

Sounds good to me.

bdraco commented:
> @bdraco Looking forward, the one piece I haven't looked into very much is that the current scan_interval configuration option is probably better in an options flow, can the configuration of configuration.yaml be used to set options as well? I thought I would bridge this during the config flow PR, but here we are.

Seems like it would be better in an options flow, how about doing the config flow now and adding the options flow later in order to keep the PR as small as possible without making this a breaking change.

MatthewFlamm commented:
Thanks @bdraco I didn't know about this possibility.  Sorry for my previous comments, needed a small amount of venting over the slow pace of review, but shouldn't have directed it at you even if I didn't intend to.

I want to think this through.  Does this strategy still make sense if there are other platforms that can be setup with the same configuration?  The functionality is all there to add NWS alert sensors in the future using the same config as here.  I should have put this in the description of the change.

It also seems to me that there are still breaking changes as I've gotten rid of a name configuration parameter, and thus the entities are also renamed.  The way it is now only requires this single breaking change.

Ideally, I would have set it up this way from the beginning but I didn't.

bdraco commented:
> Thanks @bdraco I didn't know about this possibility. Sorry for my previous comments, needed a small amount of venting over the slow pace of review, but shouldn't have directed it at you even if I didn't intend to.
> 
> I want to think this through. Does this strategy still make sense if there are other platforms that can be setup with the same configuration? The functionality is all there to add NWS alert sensors in the future using the same config as here. I should have put this in the description of the change.

If it makes sense, another potential route is adding all the sensors and disable the ones that are not popular by default. https://developers.home-assistant.io/docs/entity_registry_disabled_by/#integrations-setting-default-value-of-disabled_by-for-new-entity-registry-entries


> 
> It also seems to me that there are still breaking changes as I've gotten rid of a name configuration parameter, and thus the entities are also renamed. The way it is now only requires this single breaking change.

You could allow the name to only come in when importing from yaml to keep this from being a breaking change.  

> 
> Ideally, I would have set it up this way from the beginning but I didn't.

MatthewFlamm commented:
> If it makes sense, another potential route is adding all the sensors

This is what this PR starts doing.  Previously you either had to configure a daily entity or an hourly entity.  Now both are created.  Thus this is already a breaking change.  I don't want to have yet another breaking change when adding weather alerts sensors.  I suppose both could utilize platform configs, but that seems like not a good option either.

Can it be implemented to configure either as component or as a platform and deprecate the platform config in the future?



> You could allow the name to only come in when importing from yaml to keep this from being a breaking change.

bdraco commented:
> > If it makes sense, another potential route is adding all the sensors
> 
> This is what this PR starts doing. Previously you either had to configure a daily entity or an hourly entity. Now both are created. Thus this is already a breaking change. I don't want to have yet another breaking change when adding weather alerts sensors. I suppose both could utilize platform configs, but that seems like not a good option either.
> 
> Can it be implemented to configure either as component or as a platform and deprecate the platform config in the future?

Probably take the platform import out after 2-3 versions to give users enough time to upgrade.  No sense keeping it though if you know it is going to break for most people anyways.

> 
> > You could allow the name to only come in when importing from yaml to keep this from being a breaking change.
> 
> This cannot be done currently since I'm creating two entities now instead of one. They both cannot have the same name.

Is it possible to only assign the name to the original one that was being created in previous versions to avoid the breakage?

MatthewFlamm commented:
> Is it possible to only assign the name to the original one that was being created in previous versions to avoid the breakage?

In principle, this could be done. Would this work?.  Previously you could configure either a daily or an hourly entity with or without a specified name (defaults are daily entity type and to use the station name).  I could pass the config to the entities now and if the entity matches the type in the old config, use the old name (either the default station name or the supplied name), otherwise append daily or hourly the old name.  Since the default is daily, new users would see entities like weather.kadw and weather.kadw_hourly.

Then I assume once config flow is introduced these entity names become bound to the unique ID, so even if we change the naming logic after the deprecation period there is no breaking change?  I know this doesn't work without as I tried.

MatthewFlamm commented:
Upon further testing, entities will keep their entity_id when the backend changes only when the user had overridden it.  If the user uses the default entity_id, then it gets replaced.

To me, this will still lead to a breaking change, and a very subtle one hard to communicate.

MatthewFlamm committed:
unsubscribe listener in will_remove_from_hass

MatthewFlamm committed:
initialize _unsub_listener in __init__

MatthewFlamm commented:
After the above, I don't think there is any way of avoiding any breaking change unless we keep in logic for name and mode configuration forever.  If it is preferable to have the breaking change in 2-3 releases, I can attempt the strategy you proposed.  Are there any integrations that currently setup entries from both component and platform configurations?  In my understanding so far, this is needed.

MartinHjelmare commented:
To avoid overcomplicating this, I suggest we make a breaking change now, and require all config yaml under the integration domain top level key.

MatthewFlamm commented:
@MartinHjelmare This is a question for the future.  Does it make sense to bind the entity_id to every entity with a unique_id? As far as I can tell, this only happens for entities where the user has overwritten the entity_id.

Config flows already allow integrations to modify or handle the config entries during changes to avoid breaking changes, but if these changes also affect the entity_ids that are formed this will always be a breaking change I think.  If the entity_ids are always the same, then there are more straightforward strategies to avoid breaking changes.

MartinHjelmare commented:
Sorry, I don't understand the question.

If a unique_id is set for an entity, the user can change both name and entity_id. We as developers can also change default name and entity_id without causing a breaking change, since the entity registry keeps track of entities using the unique_id.

MatthewFlamm commented:
It is most likely that I misunderstand a subtle distinction, I need to go and do a deeper dive into the entity model.  Thanks for your patience with my questions.

MatthewFlamm committed:
use async_on_remove

MatthewFlamm committed:
remove scan interval from configuration

MatthewFlamm commented:
Not sure what is going on with codecov.

I removed scan interval configuration option, since I am mostly done the config flow import.  It wasn't obvious to me how to include it for options flow and I want to avoid the a possible breaking change.

MartinHjelmare commented:
Don't worry about codecov. It's flaky.

MatthewFlamm commented:
The test failure looks unrelated.

I believe this also makes this integration compliant with ADR 0007.  Should this go in the breaking change section for motivation purposes?

MatthewFlamm commented:
I am polishing up the tests for the config flow.  Would you prefer this to be pushed here or in a follow-up PR?

bdraco commented:
> I am polishing up the tests for the config flow. Would you prefer this to be pushed here or in a follow-up PR?

Best for the tests to come with the PR so we can see the change that is being tested

MatthewFlamm commented:
@bdraco sorry I wasn't clear.  I have config flow finished and only need to finish the tests for it.  I was wondering whether you'd prefer to see config flow here, as you hinted at above, or leave it for a second PR.

bdraco commented:
> @bdraco sorry I wasn't clear. I have config flow finished and only need to finish the tests for it. I was wondering whether you'd prefer to see config flow here, as you hinted at above, or leave it for a second PR.

Thanks for clarifying.  I've been reviewing too many config flow PRs lately :)

New PR please for the config flow + tests.

If this one ready for another review pass, please rebase on upstream/dev and repush, and let me know and I'll do another run though.

MatthewFlamm commented:
Codecov check still looks weird, I cannot figure out how to interpret it.  It has diffs from previous commits that I rebased on.

I think it is ready @bdraco

MatthewFlamm commented:
Codecov check still looks weird, I cannot figure out how to interpret it.  It has diffs from previous commits that I rebased on.

I think it is ready @bdraco

bdraco commented:
> Codecov check still looks weird, I cannot figure out how to interpret it. It has diffs from previous commits that I rebased on.
> 
> I think it is ready @bdraco

Ok to ignore.

bdraco commented on homeassistant/components/nws/weather.py:
```diff
@@ -165,85 +73,73 @@ def convert_condition(time, weather):
-    async_add_entities([NWSWeather(nws, mode, hass.config.units, config)], True)
+    async_add_entities(
+        [
+            NWSWeather(nws_data, "daynight", hass.config.units),
```
Ideally the PyPi package provides these constants.  No need to change now though -- there are a lot here so good for a followup
```suggestion
            NWSWeather(nws_data, SOME_CONSTANT_FOR_DAYLIGHT, hass.config.units),
```

MatthewFlamm replied:
This is a home assistant specific thing IMO.  The upstream pypi library just provides methods and attributes to retrieve the hourly and day+night forecasts.  Each entity needs to know which forecast it is getting, which is why I think this is a home assistant feature.  I should make these true constants either in this module or const though.

bdraco replied:
Probably a good thing to do in a followup since you have the config flow coming

bdraco commented on homeassistant/components/nws/weather.py:
```diff
@@ -253,7 +149,7 @@ def attribution(self):
     def name(self):
         """Return the name of the station."""
-        return self.station_name
+        return f"{self.station}_{self.mode}"
```
```suggestion
        return f"{self.station} {self.mode.title()}"
```

The naming looks a bit off.  Suggest the above.

<img width="455" alt="Screen Shot 2020-04-07 at 6 56 36 PM" src="https://user-images.githubusercontent.com/663432/78730141-916d2980-7901-11ea-8dc3-318f6385439f.png">

MatthewFlamm replied:
Once I wrote this out your suggestion, it reads and looks much better.

bdraco commented:
@MatthewFlamm  The naming is the only thing that really stands out that isn't an existing issue.  I'm good to approve once that is sorted.  No need to do the constants here.

MatthewFlamm committed:
Use better name

Co-Authored-By: J. Nick Koston <nick@koston.org>

MatthewFlamm commented:
Thanks @bdraco for a thorough review!  All of your comments were very helpful.

bdraco commented:
@MatthewFlamm If the CI looks good, I'll wait a day for this to show any issues on my test machine and for other reviewers to see anything I missed, and then I'll merge if it all still looks good.

bdraco commented:
LGTM (well not the weather)

<img width="475" alt="Screen Shot 2020-04-07 at 8 18 19 PM" src="https://user-images.githubusercontent.com/663432/78734103-f2e6c580-790c-11ea-9654-2db579ac6bfc.png">

bdraco commented:
No errors in the log, still working as expected.
LGTM

MartinHjelmare commented on homeassistant/components/nws/weather.py:
```diff
@@ -165,85 +73,73 @@ def convert_condition(time, weather):
+async def async_setup_platform(hass, config, async_add_entities, discovery_info):
     """Set up the NWS weather platform."""
-
     latitude = config.get(CONF_LATITUDE, hass.config.latitude)
```
We normally add a guard clause here that checks if discovery_info is None and return if so. This can happen if a user configures the platform config section by mistake.

MartinHjelmare commented on homeassistant/components/nws/weather.py:
```diff
@@ -165,85 +73,73 @@ def convert_condition(time, weather):
-        _LOGGER.debug("Forecast: %s", self._forecast)
-        _LOGGER.debug("Finished updating")
+
+        self.async_schedule_update_ha_state()
```
This should be `self.async_write_ha_state` since we don't need to fetch and update data.

MartinHjelmare commented on tests/components/nws/test_init.py:
```diff
@@ -0,0 +1,67 @@
+async def test_no_config(hass, mock_simple_nws):
+    """Test that nws does not setup with no config."""
+    with assert_setup_component(0):
+        assert await async_setup_component(hass, DOMAIN, {}) is True
```
Remove `is True`.

MartinHjelmare commented on tests/components/nws/test_init.py:
```diff
@@ -0,0 +1,67 @@
+    """Test that nws does not setup with no config."""
+    with assert_setup_component(0):
+        assert await async_setup_component(hass, DOMAIN, {}) is True
+        assert DOMAIN not in hass.data
```
It would be better to assert that there were no api calls and/or no states created. We don't want to assert integration details that might change.

MartinHjelmare reviewed (commented):
Looks good! Some comments for the future.

MatthewFlamm commented:
I will address these comments in the next PR for config flow.
